src/rate/index.tsx

```diff
@@ -0,0 +1,336 @@
+import * as iconCss from '../theme/default/icon.m.css';
+import * as radioCss from '../theme/default/radio.m.css';
+import * as radioGroupCss from '../theme/default/radio-group.m.css';
+import * as css from '../theme/default/rate.m.css';
+import bundle from './nls/Rate';
+
+export interface MixedNumber {
+	value: number;
+	quotient: number;
+	numerator: number;
+	denominator: number;
+}
+
+export interface RateProperties {
+	/** The name attribute for this rating group */
+	name: string;
+	/** The label to be displayed in the legend */
+	label?: string;
+	/** The initial rating value */
+	initialValue?: number;
+	/** Callback fired when the rating value changes */
+	onValue?: (value?: number) => void;
+	/** The number of times each star is divided */
+	steps?: number;
+	/** The highest rating for this rating group */
+	max?: number;
+	/** Allow undefined to be chosen with repeat click or with arrows */
+	allowClear?: boolean;
+	/** Prevent interaction with a visual indication */
+	disabled?: boolean;
+	/** Prevent interaction */
+	readOnly?: boolean;
+}
+
+export interface RateChildren {
+	(fill: boolean, integer: number, selected?: number, over?: number): RenderResult;
+}
+
+interface RateState {
+	uuid: string;
+	hover: string;
+	hovering: boolean;
+	selectedInteger?: number;
+	selectedStep?: number;
+	focused: boolean;
+}
+
+const icache = createICacheMiddleware<RateState>();
+
+function mixedNumber(integer: number, step: number, steps: number): MixedNumber {
+	if (step === steps) {
+		step = 0;
+	} else {
+		integer -= 1;
+	}
+	return {
+		value: integer + step / steps,
+		quotient: integer,
+		numerator: step,
+		denominator: steps
+	};
+}
+
+type RadioChecked = (checked?: boolean) => boolean | undefined;
+
+const factory = create({ focus, i18n, icache, theme })
+	.properties<RateProperties>()
+	.children<RateChildren | undefined>();
+
+export const Rate = factory(function Rate({
+	id,
+	children,
+	properties,
+	middleware: { focus, i18n, icache, theme }
+}) {
+	const { format } = i18n.localize(bundle);
+	const themedCss = theme.classes(css);
+	const {
+		name,
+		label,
+		initialValue: initialNumber = undefined,
+		onValue,
+		max = 5,
+		steps: stepsLength = 1,
+		allowClear = false,
+		disabled = false,
+		readOnly = false
+	} = properties();
+	const interaction = !disabled && !readOnly;
+	const initialStep = initialNumber ? Math.round((initialNumber % 1) * stepsLength) : undefined;
+	const initialInteger = initialNumber ? Math.ceil(initialNumber) : undefined;
+	const initialValue = initialNumber
+		? `${initialInteger}-${initialStep || stepsLength}`
+		: undefined;
+	const character = children()[0];
+	const hovering = icache.getOrSet('hovering', false);
+	const selectedInteger = icache.getOrSet('selectedInteger', initialInteger);
+	const selectedStep = icache.getOrSet('selectedStep', initialStep);
+	const shouldFocus = hovering ? focus.shouldFocus() : icache.getOrSet('focused', false);
```

Comment:
I think that should focus shouldn't be concerned with the hover status nor should it be set in the icache, just passed directly to the radio group.

Reply:
If the user is currently hovering, I use the focus middleware to temporarily show which star is active. On the next render, assuming they're still hovering, this is gone. The icache is used to notify me as focus changes between the radios (e.g. with keyboard navigation).